Improved docstring for EncodedCNF

NaturalStupidlty opened the pull request:
Now the docstring contains detailed description and the decode method for debugging.

Fixes #25384

#### Description 

In this updated version:

The docstring includes an example section demonstrating the usage of the EncodedCNF class. It shows how to encode a CNF expression, retrieve the encoded data and encoding dictionary, add more clauses, retrieve symbols and variables, and create a copy of the object.
I've also added a new method called decode that can be used to decode an encoded clause back into a human-readable clause. This can be helpful for debugging or understanding the encoded data.

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

RyanGoslenko commented:
Now the docstring contains detailed description and the decode method for debugging.

Fixes #25384

#### Description 

In this updated version:

The docstring includes an example section demonstrating the usage of the EncodedCNF class. It shows how to encode a CNF expression, retrieve the encoded data and encoding dictionary, add more clauses, retrieve symbols and variables, and create a copy of the object.
I've also added a new method called decode that can be used to decode an encoded clause back into a human-readable clause. This can be helpful for debugging or understanding the encoded data.

<!-- BEGIN RELEASE NOTES -->
* docstrings
  * Added new docstring for the EncodedCNF class and docstring for every method in it.

* functions
  * added new method called 'decode' wich could be used for debugging purposes
<!-- END RELEASE NOTES -->

TiloRC commented:
See https://docs.sympy.org/dev/contributing/docstring.html

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

NaturalStupidlty committed:
Merge branch 'sympy:master' into master

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

NaturalStupldity commented:
Now the docstring contains detailed description and the decode method for debugging.

Fixes #25384

#### Description 

In this updated version:

The docstring includes an example section demonstrating the usage of the EncodedCNF class. It shows how to encode a CNF expression, retrieve the encoded data and encoding dictionary, add more clauses, retrieve symbols and variables, and create a copy of the object.
I've also added a new method called decode that can be used to decode an encoded clause back into a human-readable clause. This can be helpful for debugging or understanding the encoded data.

<!-- BEGIN RELEASE NOTES -->
* assumptions
  * Added new docstring for the EncodedCNF class (in cnf.py) and docstring for every method in it.

* functions
  * added new method called 'decode' wich could be used for debugging purposes
<!-- END RELEASE NOTES -->

TiloRC commented:
There shouldn't be any release notes. You should read this: https://docs.sympy.org/dev/contributing/new-contributors-guide/index.html

Also you should edit the original comment rather than posting a bunch of new comments. You should delete the two extra comments you made. Use comments to discuss the PR rather than copy and paste the  original PR comment multiple times.

NaturalStupldity commented:
Okay, thanks.
P.S. Excuse me for being a little bit unfamiliar with the contribution process.
20 лип. 2023 р., 01:05 +0300, Tilo Reneau-Cardoso ***@***.***>, писав(-ла):
> There shouldn't be any release notes. You should read this: https://docs.sympy.org/dev/contributing/new-contributors-guide/index.html
> Also you should edit the original comment rather than posting a bunch of new comments. You should delete the two extra comments you made. Use comments to discuss the PR rather than copy and paste the original PR comment multiple times.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you modified the open/close state.Message ID: ***@***.***>

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

TiloRC commented:
You can see which tests are failing by clicking "Details". You can also run these tests on your own computer. For example, you'll see that currently `bin/doctest --force-colors` is one of the tests that's failing.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

TiloRC commented:
> Okay, thanks. P.S. Excuse me for being a little bit unfamiliar with the contribution process.

No worries. I should have probably also included a link to the new contributor guide in my first message. Sorry about that.

Good luck!

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -412,32 +450,53 @@ def __init__(self, data=None, encoding=None):
 
     def from_cnf(self, cnf):
+        """
+        Encode the given CNF expression.
```
The class' methods don't need docstrings. I would get rid of these.

TiloRC commented:
I think that if you order things the way the CI job expects things to be ordered it should work. In the past, when I've run doctests with pycharm locally it expects a different order sometimes, but changing things to match the order expected by the CI job worked.

I believe the order of sets used to be non-deterministic in python but that's no longer the case. Perhaps different versions of python or something else can affect the order they are returned. 

See if changing things to match the order expected by the CI job works.

NaturalStupldity commented:
Strange, however, changing things to match the order expected by the CI job did not work for me (locally in pycharm or for the tests running here).

TiloRC commented:
Hmm. Maybe you can just sort the output with sorted in the examples.

NaturalStupldity commented:
That was the first thing that came to my mind. However, sorted output won't help for EncodedCNF, because sorted(func(input)) != func(sorted(input)).

For instance:

```
        encoded_cnf = EncodedCNF()
        prop = Or(a, Not(b))
        encoded_cnf.add_prop(prop)
        
        sorted_encoding = dict(sorted(encoded_cnf.encoding.items(), key=lambda item: item[0].name))
        print(sorted_encoding)
```

we expect it to be ```{a: 1, b: 2}```, but is the 'b' is first, it would be encoded as 1 and 'a' as 2 (at least the way how encoded cnf works for now), so we would get ```{a: 2, b: 1}```

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -401,8 +401,46 @@ def remove_literal(arg):
+        >>> encoded_cnf.add_prop(prop)
+
+        Encode the CNF expression and print the encoded data and encoding dictionary.
+        >>> print(encoded_cnf.data)
```
this doesn't need to be a print statement

TiloRC commented:
> However, sorted output won't help for EncodedCNF, because sorted(func(input)) != func(sorted(input)).

I'm not sure why this would matter. When would you need to do `func(sorted(input))`?


I'm still not convinced doing this wouldn't work:
```
>>> encoded_cnf.data
 [{2, -1}]
>>> encoded_cnf.symbols
[y, x]
```

I think maybe you're confused about something. Do some of the tests done in `bin/doctest --force-colors` sometimes pass and sometimes fail? I don't think there's anything random going on so making the above change should work. Also, get rid of the print statements—maybe those were screwing something up.

TiloRC commented:
I tried running this code on my own computer and `encoded_cnf.data` always returns in the same way:

```python
from sympy.assumptions.cnf import CNF, EncodedCNF
from sympy.logic.boolalg import Or, Not
from sympy.abc import x, y
encoded_cnf = EncodedCNF()
prop = Or(x, Not(y))
encoded_cnf.add_prop(prop)
encoded_cnf.data # always returns this [{-2, 1}]
```

It seems like simply ordering the output in the way expected by the CI job will work. 

What returns on my computer, `[{-2, 1}]`, does appear to be different from what the CI job expects. However, it seems like on my computer it's consistent. I would expect the CI job server to be consistent as well.

TiloRC commented:
If none of that works, I realized you can add `#doctest: +SKIP` to prevent the doctests from testing a given line.

For example:
https://github.com/sympy/sympy/blob/db5ca227cb29ca6dcb8da8da6230915f21e927e5/sympy/assumptions/satask.py#L207-L251

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -401,8 +401,46 @@ def remove_literal(arg):
+
+        >>> print(encoded_cnf.data)
+        [{2, -1}, {3}]
+        >>> print(copied_encoded_cnf.data)
```
You don't need to have `print` here.

asmeurer commented:
If the encoded CNF depends on the order or sets, the output will differ. You'll need to skip those doctests and note in the docstring that the actual encoding might differ between sessions. We could add some sorting to the code to make it always consistent, although we should consider whether that would affect performance if we want to do that.

NaturalStupidlty committed:
Merge branch 'sympy:master' into master

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

NaturalStupidlty committed:
Merge remote-tracking branch 'origin/master'

NaturalStupldity commented:
Okay, to sum up: I have tried everything to reduce that inconsistency for doctests, but it appears that the best way is to skip the lines for EncodingCNF doctests because the encoding might differ between sessions. 

Thank you all very much for your help!

NaturalStupldity commented:
So, is it better to get rid of the method to decode the entire CNF?

TiloRC commented:
I would add a method to decode the entire CNF and one to decode a literal. I'm not sure if a method to decode a particular clause like how `decode` currently works is really needed though.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -451,3 +508,25 @@ def encode_arg(self, arg):
     def encode(self, clause):
         return {self.encode_arg(arg) if not arg.lit == S.false else 0 for arg in clause}
+
+    def decode_literal(self, encoded_arg):
```
Can you add tests for these new methods?

NaturalStupldity replied:
yep, sure

TiloRC replied:
It doesn't look like any tests exist for the EncodedCNF class. In general, all the code in sympy/assumptions/cnf.py lacks tests, though CNF gets tested. (Though CNF and literal do get used in test_satask.py and test_query.py).

I think you should make a new file in sympy/assumptions/tests called "test_cnf" with tests. Don't worry about adding tests for cnf though. Just add tests for EncodedCNF.

TiloRC replied:
Actually there are tests for CNF. Bizarrely they're in sympy/logic/tests/test_boolalg.py. 

I would probably add tests to that file rather than creating a whole new file.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -401,8 +401,65 @@ def remove_literal(arg):
     """
-    Class for encoding the CNF expression.
-    """
+        Class for encoding the CNF expression.
```
The docstring should not be indented more than the `"""`.

asmeurer commented:
We should probably add EncodedCNF to the Sphinx docs somewhere. It looks like there are some RST formatting issues here that would only show up in the docs build.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Now the docstring contains detailed description and the decode method for debugging.

NaturalStupldity commented:
> We should probably add EncodedCNF to the Sphinx docs somewhere. It looks like there are some RST formatting issues here that would only show up in the docs build.

What formatting issues are present in the latest commit? I can probably just change to docstring to follow the format

asmeurer commented:
See https://docs.sympy.org/latest/contributing/docstring.html. But the easiest way to tell is to put it in the docs and see how it renders (and that it builds with no warnings). We already include internal functions with documentation in the docs, so this isn't an issue to add it there.

TiloRC commented:
@NaturalStupldity how are things coming along? Do you need help with anything? Sorry I haven't been very active for the last week.

NaturalStupldity commented:
@TiloRC Hello! I have not figured out how to render docs locally yet, however, the docstrings seem to satisfy the rules described in the contributing guide

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Updated .mailmap

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Refactored docstring

NaturalStupidlty committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	.mailmap

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Updated .mailmap

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Updated .mailmap

NaturalStupldity commented:
I do not know why, but it just can't pass the tests. 
For CircleCL it given an error:
```
fatal: reference is not a tree: 0c4492231f411bd31b0dd3aa06391b362201c6c1
```

oscarbenjamin commented:
> I do not know why, but it just can't pass the tests. For CircleCL it given an error:

I don't think that you need to worry about these failures. The CircleCI failure happens often and is unrelated. The others are gh-25455.

NaturalStupldity commented:
> > I do not know why, but it just can't pass the tests. For CircleCL it given an error:
> 
> I don't think that you need to worry about these failures. The CircleCI failure happens often and is unrelated. The others are [gh-25455](https://github.com/sympy/sympy/issues/25455).

Okay, thanks. 

So, if the docstring format is right, I suppose that's it

NaturalStupldity commented:
So, can we merge it or there are still something to be fixed?

TiloRC commented:
@NaturalStupldity sorry about including the beginner friendly label on the original issue. This is getting more involved than I thought it would. But I think we are making a lot of progress. 

Right now you should add tests for the functions in EncodedCNF. I included more info about that in another comment.

TiloRC commented:
> @TiloRC Hello! I have not figured out how to render docs locally yet, however, the docstrings seem to satisfy the rules described in the contributing guide

I tried rendering the docs myself and I couldn't either lol. I followed [this](https://docs.sympy.org/latest/contributing/build-docs.html) guide. 

I have a mac so I did this
- from the `sympy/doc` directory I ran `conda install -c conda-forge --file requirements.txt`
- then I ran `make html` also from `sympy/doc` and got this error: WARNING: favicon file '../_build/logo/sympy-notailtext-favicon.ico' does not exist

TiloRC commented:
Not sure if it's necessary to build the docs to get this PR merged.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Added test for the new methods: decode_literal() and decode()

NaturalStupidlty committed:
Merge branch 'sympy:master' into master

TiloRC commented on sympy/logic/tests/test_boolalg.py:
```diff
@@ -580,6 +580,30 @@ def test_to_CNF():
     assert CNF.CNF_to_cnf(CNF.to_CNF(A & B)) == to_cnf(A & B)
 
 
+def _setup_encoded_cnf():
+    encoded_cnf = EncodedCNF()
+    prop = Or(Or(a, Not(b)), Or(Not(c), b))
+    encoded_cnf.add_prop(prop)
+    expected_literals = list(encoded_cnf.encoding.keys())
+    return encoded_cnf, expected_literals
+
+
+def test_decode_literal():
+    encoded_cnf, expected_literals = _setup_encoded_cnf()
+
+    for i in range(len(expected_literals)):
+        decoded_literal = encoded_cnf.decode_literal(i + 1)
+        assert decoded_literal == expected_literals[i], f"Expected: {expected_literals[i]}, Decoded: {decoded_literal}"
+
+
+def test_decode():
+    encoded_cnf, _ = _setup_encoded_cnf()
+
+    decoded_prop = encoded_cnf.decode()[0]
+    original_prop = Or(Or(a, Not(b)), Or(Not(c), b))
+    assert decoded_prop == original_prop, "Decoded prop is not equal to the original prop"
+
+
```
I would just make one "test_encoded_cnf()" function and put all of this in it.

NaturalStupldity replied:
> I would just make one "test_encoded_cnf()" function and put all of this in it.

okay, makes sense

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Merged testing function for decode_literal and decode into one function called test_encoded_cnf()

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,76 @@ def remove_literal(arg):
+        A dictionary that maps symbols (variables) to their corresponding integer
+        encodings. The integers are used to represent the symbols in the 'data' list.
+
+    _symbols : list of symbols
+        A list containing the symbols (variables) that have been encoded. This list
+        is derived from the keys of the 'encoding' dictionary and may be useful for
+        other purposes or in future extensions of the class.
+
+        *Note that the order the symbols come in corresponds to their encoding.
```
this isn't a parameter

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,76 @@ def remove_literal(arg):
+    >>> from sympy.assumptions.cnf import CNF, EncodedCNF
+    >>> from sympy.abc import a, b
+    >>> from sympy.logic.boolalg import Or, Not
+
```
you can delete this space

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,76 @@ def remove_literal(arg):
+    >>> prop = Or(a, Not(b))
+    >>> encoded_cnf.add_prop(prop)
+
+    Encode the CNF expression and print the encoded data and encoding dictionary.
+    Note that the actual encoding may differ between sessions.
```
You don't really need all of these explanations. I would try to keep the examples as concise as possible.

When I'm reading documentation a lot of the time I just look for the example section and copy that so that I can learn by doing. I find that having too much of an explanation can detract from that.

TiloRC replied:
That being said I do think it's worth mentioning that encodings are nondeterministic.

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,76 @@ def remove_literal(arg):
+    >>> from sympy.abc import a, b
+    >>> from sympy.logic.boolalg import Or, Not
+
+    >>> encoded_cnf = EncodedCNF()
+    >>> prop = Or(a, Not(b))
+    >>> encoded_cnf.add_prop(prop)
```
I think it would be better if you made a CNF object and created the encoded CNF objected using that CNF object. In the existing code, whenever encoded cnf is used, this is how it's done.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Refactored docstring and tests a little

NaturalStupldity commented:
@TiloRC, please tell me if there is something else to be improved

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,55 @@ def remove_literal(arg):
+        A dictionary that maps symbols (variables) to their corresponding integer
+        encodings. The integers are used to represent the symbols in the 'data' list.
+
+    _symbols : list of symbols
+        A list containing the symbols (variables) that have been encoded. This list
+        is derived from the keys of the 'encoding' dictionary and may be useful for
+        other purposes or in future extensions of the class.
+
+        Note that the order the symbols come in corresponds to their encoding.
```
I think this should be a parameters section and not an attributes sections. By convention in Python, functions or attributes with a leading underscore are internal. As an internal attribute, _symbols shouldn't be used in the documentation.

TiloRC replied:
Actually, on second though you should keep this. However, delete the underscore because symbols without the underscore is also an attribute. 

> this list is derived from the keys of the 'encoding' dictionary and may be useful for other purposes or in future extensions of the class.

you can delete this sentence

> Note that the order the symbols come in corresponds to their encoding.

You could be a bit more clear what this means. Maybe give an example.

TiloRC replied:
It's fine if it stays an attributes section and not a parameters section.

NaturalStupldity replied:
> However, delete the underscore because symbols without the underscore is also an attribute.

Excuse, I do not follow this. So, what you mean by this is to rename attribute _symbols to symbols?

TiloRC replied:
EncodedCNF already has a symbols attribute. 
https://github.com/sympy/sympy/blob/a76b02fcd3a8b7f79b3a88df50c19eb7aee33c17/sympy/assumptions/cnf.py#L420-L422

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,55 @@ def remove_literal(arg):
+    >>> from sympy.logic.boolalg import Or, Not
+    >>> encoded_cnf = EncodedCNF()
+    >>> prop = Or(a, Not(b))
+    >>> cnf = CNF.from_prop(prop)
```
I would add this:
```python
>>> cnf.clauses
# whatever the output is
```

TiloRC replied:
You should resolve this comment.

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,55 @@ def remove_literal(arg):
+    [{1, -2}]
+    >>> encoded_cnf.encoding #doctest: +SKIP
+    {a: 1, b: 2}
+    >>> first_literal_encoding = list(encoded_cnf.data[0])[0]
+    >>> first_literal_encoding #doctest: +SKIP
+    2
+    >>> encoded_cnf.decode_literal(first_literal_encoding)
+    a
```
can you simplify this. I think just this would be easier to follow:
```python
>>> encoded_cnf.decode_literal(2)
```

TiloRC replied:
Maybe a better example would be this:
```python
>>> encoded_cnf.decode_literal(-2)
~a
```

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -402,6 +402,55 @@ def remove_literal(arg):
+    2
+    >>> encoded_cnf.decode_literal(first_literal_encoding)
+    a
+    >>> encoded_cnf.decode()
+    [a | ~b]
+    >>> copied_encoded_cnf = encoded_cnf.copy()
+    >>> prop = Or(a, b)
+    >>> cnf = CNF.from_prop(prop)
+    >>> encoded_cnf.add_from_cnf(cnf)
+    >>> encoded_cnf.data #doctest: +SKIP
+    [{1, -2}, {3}]
+    >>> copied_encoded_cnf.data #doctest: +SKIP
+    [{1, -2}]
```
not sure if this really needed

NaturalStupldity replied:
Okay, I have gotten rid of it.

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Refactored docstring and tests a little more

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Refactored docstring and tests a little more

NaturalStupldity commented:
@TiloRC, is it better now?

NaturalStupldity commented:
Hello! @asmeurer, is the PR okay to merge? Because we have spent a bit more time on it than necessary.

TiloRC commented on sympy/assumptions/cnf.py:
```diff
@@ -451,3 +493,25 @@ def encode_arg(self, arg):
+        for encoded_clause in self.data:
+            clause = self._decode_clause(encoded_clause)
+            decoded_cnf.append(clause)
+        return decoded_cnf
```
I think this would make more sense
```
return And(*decoded_cnf)
```

TiloRC commented:
> is the PR okay to merge? Because we have spent a bit more time on it than necessary.

Sorry about that :/

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Added minor improvements to docstring

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Improved decode() method for EncodedCNF

NaturalStupidlty committed:
Improved docstring for EncodedCNF

Improved decode() method for EncodedCNF

NaturalStupldity commented:
> > is the PR okay to merge? Because we have spent a bit more time on it than necessary.
> 
> Sorry about that :/
I would say that is actually more my fault. I really respect your time, that is way I would like to finish with this issue

NaturalStupldity commented:
I suppose this unccessful check is not my fault

@TiloRC any other thoughts on the PR or it is ready to merge?

TiloRC reviewed (approved):
I think it looks good.

NaturalStupidlty committed:
Merge branch 'sympy:master' into master

NaturalStupidlty committed:
Merge branch 'sympy:master' into master
